pkg/controller/iam/managed.go

```diff
@@ -0,0 +1,261 @@
+	populateProviderFromCR(sa, cr)
+	psar := &iamv1.PatchServiceAccountRequest{
+		ServiceAccount: sa,
+		UpdateMask:     "description,displayName",
+	}
+	req := e.serviceAccounts.Patch(e.rrn.ResourceName(cr), psar)
+	// we don't pay attention to the result of the patch request because it is only guaranteed to contain
+	// `description` and `displayName` ie the fields we are trying to change
+	_, err := req.Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
+	}
+	return managed.ExternalUpdate{}, nil
+}
+
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/delete
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return errors.New(errNotServiceAccount)
+	}
+
+	req := e.serviceAccounts.Delete(e.rrn.ResourceName(cr))
+	_, err := req.Context(ctx).Do()
+
+	if err != nil && gcp.IsErrorNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, errDelete)
+}
+
+// isUpToDate returns true if the supplied Kubernetes resource does not differ
+//  from the supplied GCP resource. It considers only fields that can be
+//  modified in place without deleting and recreating the Service Account.
+func isUpToDate(in *v1beta1.ServiceAccountParameters, observed *iamv1.ServiceAccount) bool {
+	// see comment in serviceaccount_types.go
+	if in.DisplayName != nil && *in.DisplayName != observed.DisplayName {
+		return false
+	}
+	if in.Description != nil && *in.Description != observed.Description {
+		return false
+	}
+	return true
+}
+
+func populateCRFromProvider(cr *v1beta1.ServiceAccount, fromProvider *iamv1.ServiceAccount) {
+	cr.Status.AtProvider.UniqueID = fromProvider.UniqueId
+	cr.Status.AtProvider.Email = fromProvider.Email
+	cr.Status.AtProvider.Oauth2ClientID = fromProvider.Oauth2ClientId
+	cr.Status.AtProvider.Disabled = fromProvider.Disabled
+	cr.Status.AtProvider.Name = fromProvider.Name
+}
+
+func populateProviderFromCR(forProvider *iamv1.ServiceAccount, cr *v1beta1.ServiceAccount) {
+	forProvider.DisplayName = gcp.StringValue(cr.Spec.ForProvider.DisplayName)
+	forProvider.Description = gcp.StringValue(cr.Spec.ForProvider.Description)
+}
+
+// NewRelativeResourceNamer makes an instance of the RelativeResourceNamer
+// which is the only type that is allowed to know how to contsruct GCP resource names
+// for the IAM type.
+func NewRelativeResourceNamer(projectName string) RelativeResourceNamer {
+	return RelativeResourceNamer{projectName: projectName}
+}
+
+// RelativeResourceNamer allows the controller to generate the "relative resource name"
+// for the service account and GCP project based on the external-name annotation.
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+// The relative resource name for service accounts has the following format:
+// projects/{project_id}/serviceAccounts/{account name}
+type RelativeResourceNamer struct {
```

Comment:
Here's my reasoning. A manager exists in the scope of a project; project ID seems equivalent to things like Connectors that we reuse between reconciliations. I like the Observe/Create/Update/Delete methods being somewhat unconcerned with the details of how the project ID is used in constructing the Relative Resource Name, because that's a funny detail specific to how GCP works. I don't like that we have to plumb the projectID around into different functions, hiding it in this type removes that concern and makes it very clear that the only reason these methods need to know the project ID is to construct these resource names, which they delegate to the type. I think it's pretty clear from the `New` function and other attributes of the code how the project name will be used.